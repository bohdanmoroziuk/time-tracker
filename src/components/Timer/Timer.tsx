import { FunctionComponent } from 'react';
import { Text, View } from 'react-native';

import { millisecondsToHuman } from 'src/utils/timer';
import TimerButton from 'src/components/TimerButton';

import styles from './Timer.styles';
export interface TimerProps {
  id: string;
  title: string;
  project: string;
  elapsed: number;
  isRunning?: boolean;
}

const Timer: FunctionComponent<TimerProps> = ({
  title,
  project,
  elapsed
}) => {
  const elapsedTime = millisecondsToHuman(elapsed);

  return (
    <View style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      <Text>{project}</Text>
      <Text style={styles.elapsedTime}>{elapsedTime}</Text>
      <View style={styles.buttonGroup}>
        <TimerButton small color="blue" title="Edit" onPress={() => {}} />
        <TimerButton small color="red" title="Remove" onPress={() => {}} />
      </View>
      <TimerButton color="#21BA45" title="Start" onPress={() => {}} />
    </View>
  );
};

export default Timer;