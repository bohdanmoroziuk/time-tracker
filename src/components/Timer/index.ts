import Timer from './Timer';

export type { TimerProps } from './Timer';
export default Timer;
