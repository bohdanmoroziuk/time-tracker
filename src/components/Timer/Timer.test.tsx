import { render, fireEvent } from '@testing-library/react-native';
import { Alert } from 'react-native';

import Timer from './Timer';

jest.spyOn(Alert, 'alert');

const mockRemoveTimer = jest.fn();
const mockResetTimer = jest.fn();
const mockToggleTimer = jest.fn();

jest.mock('../../contexts/timers', () => ({
  useTimers: jest.fn().mockImplementation(() => ({
    removeTimer: mockRemoveTimer,
    resetTimer: mockResetTimer,
    toggleTimer: mockToggleTimer,
  })),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Timer', () => {
  const id = '1';
  const title = 'Abs';
  const project = 'Workout';
  const elapsed = 132323;
  const humanReadableElapsed = "00:02:12";
  const isRunning = false;
  const mockOnEdit = jest.fn();

  it('renders successfully', () => {
    const { toJSON } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const { getByText } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    expect(getByText(title)).toBeDefined();
    expect(getByText(project)).toBeDefined();
    expect(getByText(humanReadableElapsed)).toBeDefined();
  });

  it('enables editing mode', () => {
    const { getByText } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    fireEvent.press(getByText('Edit'));

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
  });

  it('removes the timer', () => {
    const { getByText } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    fireEvent.press(getByText('Remove'));

    /**
     * @see https://stackoverflow.com/a/61125322
     * @see https://spectrum.chat/testing-library/help-react-native/how-do-you-test-an-alert~c05e61a9-c46d-4e2b-b012-3d5d2497aa79?m=MTU3NDA5NjQ0NDMzOQ==
     */
    (Alert.alert as jest.Mock).mock.calls[0][2][1].onPress();

    expect(mockRemoveTimer).toHaveBeenCalledTimes(1);
    expect(mockRemoveTimer).toHaveBeenCalledWith(id);
  });

  it('resets the timer', () => {
    const { getByText } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    fireEvent.press(getByText('Reset'));

    expect(mockResetTimer).toHaveBeenCalledTimes(1);
    expect(mockResetTimer).toHaveBeenCalledWith(id);
  });

  it('toggles the timer state', () => {
    const { getByText, rerender } = render(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning,
          onEdit: mockOnEdit,
        }}
      />
    );

    fireEvent.press(getByText('Start'));

    expect(mockToggleTimer).toHaveBeenCalledTimes(1);
    expect(mockToggleTimer).toHaveBeenCalledWith(id);

    rerender(
      <Timer
        {...{
          id,
          title,
          project,
          elapsed,
          isRunning: true,
          onEdit: mockOnEdit,
        }}
      />
    );

    fireEvent.press(getByText('Stop'));

    expect(mockToggleTimer).toHaveBeenCalledTimes(2);
    expect(mockToggleTimer).toHaveBeenCalledWith(id);
  });
});
