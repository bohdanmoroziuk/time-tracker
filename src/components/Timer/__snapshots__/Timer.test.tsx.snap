// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer renders successfully 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderColor": "#d6d7da",
      "borderRadius": 10,
      "borderWidth": 2,
      "margin": 15,
      "marginBottom": 0,
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 14,
        "fontWeight": "bold",
      }
    }
  >
    Abs
  </Text>
  <Text>
    Workout
  </Text>
  <Text
    style={
      Object {
        "fontSize": 26,
        "fontWeight": "bold",
        "paddingVertical": 15,
        "textAlign": "center",
      }
    }
  >
    00:02:12
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "blue",
          "borderRadius": 3,
          "borderWidth": 2,
          "marginTop": 10,
          "minWidth": 100,
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "fontSize": 14,
              "padding": 5,
            },
            Object {
              "color": "blue",
            },
          ]
        }
      >
        Edit
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "black",
          "borderRadius": 3,
          "borderWidth": 2,
          "marginTop": 10,
          "minWidth": 100,
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "fontSize": 14,
              "padding": 5,
            },
            Object {
              "color": "black",
            },
          ]
        }
      >
        Reset
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "red",
          "borderRadius": 3,
          "borderWidth": 2,
          "marginTop": 10,
          "minWidth": 100,
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "fontSize": 14,
              "padding": 5,
            },
            Object {
              "color": "red",
            },
          ]
        }
      >
        Remove
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderColor": "green",
        "borderRadius": 3,
        "borderWidth": 2,
        "marginTop": 10,
        "minWidth": 100,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "fontSize": 16,
            "padding": 10,
          },
          Object {
            "color": "green",
          },
        ]
      }
    >
      Start
    </Text>
  </View>
</View>
`;
